File: materials/ZeroJudge/i401/i401.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;
#define pb push_back

vector<vector<pair<ll,ll>>> x_based(30001);
vector<vector<pair<ll,ll>>> y_based(60001);

int main()
{
    ll n;
    cin >> n;

    for (ll i = 0; i < n; i++)
    {
        ll xx, yy, tt;
        cin >> xx >> yy >> tt;

        // 固定 x，按照 y 排列
        x_based[xx].pb({yy, tt});

        // 固定 y，按照 x 排列
        y_based[yy + 30000].pb({xx, tt});
    }

    // 一次 sort 就好
    for (auto &v : x_based)
        sort(v.begin(), v.end());

    for (auto &v : y_based)
        sort(v.begin(), v.end());


    ll ans = 0;

    char face = 'r';

    ll x = 0;
    ll y = 0;

    while (true)
    {
        // --------------------
        // right
        // --------------------
        if (face == 'r')
        {
            auto &v = y_based[y + 30000];

            ll ub = upper_bound(v.begin(), v.end(), make_pair(x, LLONG_MAX)) - v.begin();

            if (ub == v.size())
            {
                cout << ans << '\n';
                return 0;
            }

            x = v[ub].first;
            ll type = v[ub].second;

            ans++;

            if (type == 1) // "\"
                face = 'd';
            else           // "/"
                face = 'u';
        }

        // --------------------
        // left
        // --------------------
        else if (face == 'l')
        {
            auto &v = y_based[y + 30000];

            ll lb = lower_bound(v.begin(), v.end(), make_pair(x, LLONG_MIN)) - v.begin();

            if (lb == 0)
            {
                cout << ans << '\n';
                return 0;
            }

            lb--;

            x = v[lb].first;
            ll type = v[lb].second;

            ans++;

            if (type == 1) // "\"
                face = 'u';
            else           // "/"
                face = 'd';
        }

        // --------------------
        // up
        // --------------------
        else if (face == 'u')
        {
            auto &v = x_based[x];

            ll ub = upper_bound(v.begin(),v.end(),make_pair(y, LLONG_MAX)) - v.begin();

            if (ub == v.size())
            {
                cout << ans << '\n';
                return 0;
            }

            y = v[ub].first;
            ll type = v[ub].second;

            ans++;

            if (type == 1) // "\"
                face = 'l';
            else           // "/"
                face = 'r';
        }

        // --------------------
        // down
        // --------------------
        else if (face == 'd')
        {
            auto &v = x_based[x];

            ll lb = lower_bound(v.begin(), v.end(), pair<ll,ll>{y, LLONG_MIN}) - v.begin();

            if (lb == 0)
            {
                cout << ans << '\n';
                return 0;
            }

            lb--;

            y = v[lb].first;
            ll type = v[lb].second;

            ans++;

            if (type == 1)
                face = 'r';
            else
                face = 'l';
        }
    }
}

File: materials/ZeroJudge/i401/i401.py
import bisect

n = int(input())

x_based = [[] for _ in range(30001)]
y_based = [[] for _ in range(60001)]

for _ in range(n):
    xx, yy, tt = map(int, input().split())
    x_based[xx].append([yy, tt])
    y_based[yy + 30000].append([xx, tt])

for v in x_based:
    v.sort()

for v in y_based:
    v.sort()

ans = 0

x = 0
y = 0
face = 'r'

while True:

    if face == 'r':
        v = y_based[y + 30000]
        ub = bisect.bisect_right(v, [x, 1])

        if ub == len(v):
            print(ans)
            break

        x = v[ub][0]
        typ = v[ub][1]

        ans += 1

        if typ == 1:
            face = 'd'
        else:
            face = 'u'


    elif face == 'l':
        v = y_based[y + 30000]
        lb = bisect.bisect_left(v, [x, 0])

        if lb == 0:
            print(ans)
            break

        lb -= 1

        x = v[lb][0]
        typ = v[lb][1]

        ans += 1

        if typ == 1:
            face = 'u'
        else:
            face = 'd'


    elif face == 'u':
        v = x_based[x]
        ub = bisect.bisect_right(v, [y, 1])

        if ub == len(v):
            print(ans)
            break

        y = v[ub][0]
        typ = v[ub][1]

        ans += 1

        if typ == 1:
            face = 'l'
        else:
            face = 'r'


    elif face == 'd':
        v = x_based[x]
        lb = bisect.bisect_left(v, [y, 0])

        if lb == 0:
            print(ans)
            break

        lb -= 1

        y = v[lb][0]
        typ = v[lb][1]

        ans += 1

        if typ == 1:
            face = 'r'
        else:
            face = 'l'